aptos-indexer-processors-sdk/testing-framework/src/sdk_test_context.rs

```diff
@@ -0,0 +1,132 @@
+use crate::{database::new_db_pool, mock_grpc::MockGrpcServer};
+use anyhow::Context;
+use aptos_indexer_processor_sdk::{
+    aptos_indexer_transaction_stream::TransactionStreamConfig,
+    traits::processor_trait::ProcessorTrait,
+};
+use aptos_protos::{indexer::v1::TransactionsResponse, transaction::v1::Transaction};
+use diesel::{pg::PgConnection, Connection};
+use testcontainers::{
+    core::{IntoContainerPort, WaitFor},
+    runners::AsyncRunner,
+    ContainerAsync, GenericImage, ImageExt,
+};
+use url::Url;
+
+pub struct SdkTestContext {
```

Comment:
How difficult would it be to move the DB-specific setup to a trait? With SDK we've removed all references to diesel/Postgres and instead provide traits to let the processors define the DB dependency and queries. Wonder if we can abstract those parts in the testing framework as well so it could potentially work for any DB?

Reply:
I _think_ it's doable, if we move the container setup and get_db_url into a trait. And change the `run` function to return the db url instead of the connection.

Reply:
that approach should not be overly difficult, so it would look like SdkTestContext to work with any test DB. something like this
trait 
```
pub trait TestDatabase: {
    fn establish_connection(&self) -> anyhow::Result<>
}
```

```
pub struct SdkTestContext<DB: TestDatabase> {
    pub transaction_batches: Vec<Transaction>,
    pub database: DB,
}

```
i am not sure what else needs to be abstracted away and might need some insights of defining database trait to be sufficiently flexible while it's still eays to use and working.

Reply:
assume they implemented that trait for Postgre, then they can write test like 

```
#[tokio::test]
async fn test_run() {
    let imported_txns = [IMPORTED_TESTNET_TXNS_1255836496_V2_FA_METADATA_];

    let db = PostgresTestDatabase {
        connection_string: "postgres://postgres:postgres@localhost/postgres".to_string(),
    };

    let test_context = SdkTestContext::new(&imported_txns, db)
        .await
        .unwrap();

    let events_processor_config = EventsProcessorConfig {
        channel_size: 10,
        per_table_chunk_sizes: AHashMap::new(),
    };
    let processor_config = ProcessorConfig::EventsProcessor(events_processor_config);
    let events_processor = test_context
        .init_events_processor(
            db.connection_string.clone(),
            processor_config,
            Some(1255836496),
            Some(1255836496),
        )
        .await;


    test_context
        .run(
            &events_processor,
            test_type,
            move |conn, _txn_version| {
                // Example validation logic
                conn.execute_transaction("1")?;
                Ok(())
            },
        )
        .await
        .unwrap();
}


```

Reply:
run function 
```
    pub async fn run<F>(
        &self,
        processor: &impl ProcessorTrait,
        test_type: TestType,
        verification_f: F,
    ) -> anyhow::Result<()>
    where
        F: Fn(Box<dyn DbConnection>, &str) -> anyhow::Result<()> + Send + Sync + 'static,
    {

        let db_conn = self.database.establish_connection()?;

        processor.run_processor().await?;

           test_type.run_verification(&db_conn, "1", &verification_f)?;
       

        Ok(())
    }
```